### PR TITLE
Modal a11y improvements

### PR DESCRIPTION
**Pull request by zackkrida**
Hi there! Awesome way to introduce folks to css grid. I made some quick improvements to the accessibility of the modal, which can be viewed here: ~https://css-grid-generator-a11y.zackkrida.now.sh/ (sorry I used `now` hehe 😊).~ (You've got netlify's deployment stuff going on already, might as well review there)

Should satisfy most of the concerns in #5.

## Changes:
- [x] Modal close on 'esc' keypress and on click outside of modal
- [x] Autofocus "Copy code" button when modal is opened
- [x] ARIA roles and labels added to modal
- [x] Switch `alert` on code copy success to text change within the button, with aria-live attribute
- [x] Focus trap when modal is open (not sure of best way to implement w/o adding a 3rd party lib)

I'd love to have a screenreader user assess these changes, although everything should be pretty straightforward. Hope this helps!

### CODE REVIEW

**Comment by sdras** [comment]
This is amazing! you're the best. One extra thing someone mentioned, height needs to be set on mobile with an overflow-y: scroll: https://twitter.com/bologasergiu/status/1133450289422491648. I can also follow up PR with this if you've got too much in this one. THANKS!

**Comment by zackkrida** [comment]
I think that's a great call! I'd say mobile modal access still counts as a11y. I reduced the padding of the modal on mobile (so vertical scrolling shouldn't be needed on _most_ devices) and added the vertical overflow scroll for <sup>_super tiny_</sup> displays.

File: src/components/AppCode.vue [hunk 1]
```diff
@@ -70,8 +86,14 @@ export default {
         selection.addRange(range);
       }
 
-      document.execCommand("copy");
-      alert("copied!");
+      let copied = document.execCommand("copy");
+
+      if (copied) {
+        this.codeWasCopied = true;
+        setTimeout(() => {
```
Comment (sdras) [inline src/components/AppCode.vue]: Not sure about this change: why does this need to wait 2 seconds for a11y purposes?

**Comment by salomoneb** [comment]
Hey @sdras and @zackkrida , 

I was looking to make similar tweaks when I saw this PR. I can't seem to pull the branch down to make changes (is that functionality disabled on this repo or am I missing something?), but I had a suggestion about the modal esc key open/close part. 

Rather than have all this: 
```
mounted() {
  document.addEventListener("keydown", this.closeOnEsc);
},
beforeDestroy() {
  document.removeEventListener("keydown", this.closeOnEsc);
},
methods: {
  closeOnEsc(event) {
    if (event.keyCode == 27) {
      this.close();
    }
  }
}
```
Can we just replace it with:
```
<div class="modal-mask" @keyup.esc="close" tabindex="-1">
```
To solve the "focus trap when the modal is open" item that you mentioned, this seemed to work for me in testing: 
```
mounted() {
  this.$el.focus()
}
```

Small changes, but feels a little cleaner. Let me know what you think.

**Comment by sdras** [comment]
Right, yes that's much cleaner, we should be using keycode directive modifiers here.

File: src/components/AppModal.vue [hunk 2]
```diff
@@ -1,25 +1,47 @@
...
       </div>
     </div>
   </transition>
 </template>
 
 <script>
-export default {};
+export default {
+  mounted() {
+    document.addEventListener("keydown", this.closeOnEsc);
```
Comment (trickstival) [inline src/components/AppModal.vue]: I'm excited to see stuff like this implemented in the new hooks / composable components API!

File: src/components/AppCode.vue (hunk 1 above)
Comment (trickstival) [reply src/components/AppCode.vue]: @sdras I'm not the author of the PR and I don't know if it's an a11y improvement, but the text temporary text change makes sense to me. 

This way the user knows it was copied and have the opportunity to see the old "Copy Me" text 2 seconds later. 

The alert() function just freezes the page.

Comment (sdras) [reply src/components/AppCode.vue]: Sorry, I wasn't clear, yeah I understand the functionality, I was wondering why the wait needed to be 2 seconds in particular, if there's a benefit for a11y purposes or if it was an arbitrary amount of time :)

Comment (zackkrida) [reply src/components/AppCode.vue]: 2 seconds is just an arbitrary amount of time for what's essentially a 'flash message' to the user. Only necessary if they intend to copy multiple times.

I don't know if there's any research on the ideal timeout for this sort of message!

**Comment by zackkrida** [comment]
Hi @salomoneb, I like the spirit of keeping things more vue-ey but there's specific reasons for the way I wrote this:

```js
mounted() {
  this.$el.focus()
}
```

Doesn't trap focus, it just focused the modal on mount. The user can still tab out of the modal, so there's no true "trap".

In this PR I have the `autofocus` attribute on the "Copy to clipboard!". It's preferable to focus on _this_ initially instead of the modal body itself.

_However_, with the focus on the button rather than the modal body, the event listener for `esc` needs to be on the document, not on the modal element itself. Thus the `eventListener`s being added with vanilla js.

It's definitely _preferable_ to use vue's apis when possible, but in this case I think the benefits are significant enough to add a few extra lines of vanilla js.

**Comment by salomoneb** [comment]
Gotcha. Thanks for clarifying. I was curious what other people had done on and did some digging. I found a few potential approaches, all with pros and cons. One that stood out was 18F's: 

* Page - https://accessibility.18f.gov/keyboard/#keyboard-trap
* Code - https://accessibility.18f.gov/assets/js/keyboard-access.js (you could do this without jQuery)

On their page, they have a visually hidden element with the class `.moveFocus`. When that element comes into focus, it sends the focus back to the first element in the series. The trap is disabled with some other action (could be a click, whatever).

This approach does require creating an extra element, but I like it better than all of the other JS-heavy ones I found.

**Comment by zackkrida** [comment]
Alright @salomoneb you've motivated me to add the focus trap 👏 I used the `focus-trap` js lib and it only took a few lines to wire it up to the modal.

@sdras as always no rush in reviewing all the new discussion here 😆, and i hope a few things have been cleared up! I can always rebase this branch if that'd be useful too, pulling in the i18n changes will be the main bit to resolve. Cheers!

File: src/components/AppCode.vue (hunk 1 above)
Comment (sdras) [reply src/components/AppCode.vue]: Great, thank you for clarifying :)

**Comment by sdras** [comment]
Thanks for all the good work on this @zackkrida! I'll wait to review once the merge conflicts and i18n is in :)

**Comment by zackkrida** [comment]
This should be ready to test 👍

**Comment by sdras** [comment]
Thank you! You rock! 🤩

**Comment by zackkrida** [comment]
😸thanks for putting such a cool project together! happy to help.

**Pull request merged**